### PR TITLE
Add class DetermineStationTimingOffsets

### PR DESCRIPTION
**Pull request by tomkooij**
~~This is WIP, just for comments/review:~~

This implements a ```class DetermineStationTimingOffsets``` in ```sapphire/analysis/calibration.py``` that determines stations offsets from timedelta's.

The method ```read_dt()``` can be overloaded for use on publicdb: It must read timedelta's for the appropriate file for each day.

TODO:
- [x] Add tests
- [x] Remove debugging output and ```force_stale=True```
- [x] Add electronics changes to cuts
- [x] determine MAX_DISTANCE
- [x] determine and set a minium number of timedelta's to use in fit, MIN_LEN_DT (hardcoded to 100 now)
- [x] time ```get_station_pairs_with_max_distance()``` with full network: (~5000 combinations). Perhaps iterate per cluster if too slow: Not necessay; 1 second.
- [x] ```allow_missing=True``` for HiSPARCNetwork() in ```__init__()``` ? : Not neccessary.
- [x] move to issue ~~low-priority~~: I discovered a ```datetime_range``` in ```publicdb.py``` that is a bit different. Make it DRY: Move that datetime_range to ```utils.py``` and use it here. Rename the datetime_range from this branch (that spreads the remaining days over different intervals) to an appropriate name. And/or merge.

### CODE REVIEW

**Comment by 153957** [comment]
Please use `station` or `station number` for arguments the will refer to a station number. `station id` is only to be used to refer to the 'index' of a station in a list, see Cluster objects, station ids are defined by the number in which the stations are added, while the station numbers refer to the actual station numbers.

File: sapphire/analysis/calibration.py [hunk 1]
```diff
@@ -86,6 +91,67 @@ def determine_detector_timing_offset(dt, dz=0):
...
+        :param station_id: station for which to determine station offsets
+        :param ref_station_id: station id of reference station
+        """
+        self.data = data
+        self.station_id = station_id
+        self.ref_station_id = ref_station_id
+        self.station = Station(station_id)
+        self.ref_station = Station(ref_station_id)
+        self.cluster = HiSPARCStations([station_id, ref_station_id])
+        self.r, _, self.dz = self.cluster.calc_rphiz_for_stations(0, 1)
```
Comment (153957) [inline sapphire/analysis/calibration.py]: This `r` should depend on the timestamp/dates for which the offset is determined, so for each split determine `r`.

**Comment by tomkooij** [comment]
Will do! 
I need to refactor most of it anyway, but I'll focus on getting the intervals in which we determine offsets correct first.

**Comment by tomkooij** [comment]
station timing offsets determined by this PR versus the offsets I have stored as local data (topaz/api):
station_timing_offsets

(script: https://gist.github.com/tomkooij/92c99dc01586d0cdf187ebec1ed04f60)

**Comment by tomkooij** [comment]
I filled my publicdb test VM with timedelta's and added this class to ../histograms/esd.py (with some logic to actually read from esd h5 files):
station_timing_offsets_publicdb

:smiley:

**Comment by 153957** [comment]
About the `datetime_range`-type functions, I think we should move those to `time_util.py`.
I also have a `monthrange` in topaz that could be (with some modifications) placed along those..

**Comment by tomkooij** [comment]
Still WIP.

File: sapphire/analysis/calibration.py [hunk 2]
```diff
@@ -86,6 +93,196 @@ def determine_detector_timing_offset(dt, dz=0):
...
+        if stations is not None:
+            self.cluster = HiSPARCStations(stations, skip_missing=True,
+                                           force_stale=self.force_stale)
+        else:
+            self.cluster = HiSPARCNetwork(force_stale=self.force_stale)
+
+    def read_dt(self, station, ref_station, start, end):
+        """ Overload this method on publicdb """
+
+        table = self.data.get_node('/s%d' % station)
```
Comment (153957) [inline sapphire/analysis/calibration.py]: Correct path is `/time_deltas/station_{ref_station}/station_{station}/time_deltas` https://github.com/HiSPARC/sapphire/blob/master/sapphire/analysis/time_deltas.py#L151

Comment (tomkooij) [inline sapphire/analysis/calibration.py]: I changed the path. (This was the path in my testfile, it matches the one in the ProcessTimeDeltas class now).
```read_dt()``` is overloaded on publicdb, but it uses the same path

**Comment by tomkooij** [comment]
I added unittests for most of DetermineStationTimingOffsets (at least everything that is used on publicdb). There is still a lot of calibration.py without coverage, but that's not in this PR.

This seems to work on my publicdb VM.

I think this can be merged to be tested "live" on publicdb.

**Comment by 153957** [comment]
Ill just leave this here.. http://stackoverflow.com/a/35674989/1033535

**Pull request merged**